File: webForActors/src/main/java/unibo/webForActors/HIController.java
package unibo.webForActors;

import connQak.ConnQakBase;
import it.unibo.kactor.IApplMessage;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import unibo.Robots.MainBasicRobot;
import unibo.Robots.MainRobotCleaner;
import unibo.actor22.Qak22Context;
import unibo.actor22.Qak22Util;
import unibo.actor22comm.ProtocolType;
import unibo.actor22comm.SystemData;
import unibo.actor22comm.interfaces.Interaction2021;
import unibo.actor22comm.utils.ColorsOut;
import unibo.actor22comm.utils.CommSystemConfig;
import unibo.actor22comm.utils.CommUtils;

@Controller
public class HIController {
    private static final String robotCmdId = "move";
    private static  String robotName  = "";

    private Interaction2021 conn;

    public HIController(){
        ColorsOut.outappl("HIController: CREATE"   , ColorsOut.WHITE_BACKGROUND);
        //createRobotCleaner();
        createBasicRobot();
    }

    protected void createRobotCleaner(){
        robotName  = MainRobotCleaner.myName;
        CommUtils.aboutThreads("Before start - ");
        MainRobotCleaner appl = new MainRobotCleaner( );
        appl.doJob();
        //appl.terminate();
    }
    protected void createBasicRobot(){
        CommSystemConfig.tracing = true;
        robotName  = MainBasicRobot.myName;
        CommUtils.aboutThreads("Before start - ");
        MainBasicRobot appl = new MainBasicRobot( );
        appl.doJob();
        //appl.terminate();
    }
    private IApplMessage moveAril( String cmd  ) {
        //ColorsOut.outappl("HIController | moveAril cmd:" + cmd , ColorsOut.BLUE);
        switch( cmd ) {
            case "w" : return CommUtils.buildDispatch("webgui", robotCmdId, "w", robotName);
            case "s" : return CommUtils.buildDispatch("webgui", robotCmdId, "s",robotName);
            case "a" : return CommUtils.buildDispatch("webgui", robotCmdId, "a",    robotName);
            case "d" : return CommUtils.buildDispatch("webgui", robotCmdId, "d",   robotName);
            case "h" : return CommUtils.buildDispatch("webgui", robotCmdId, "h",       robotName);
            default: return CommUtils.buildDispatch("webgui",   robotCmdId, "h",       robotName);
        }
    }
    @Value("${spring.application.name}")
    String appName;

    @GetMapping("/")
    public String homePage(Model model) {
        model.addAttribute("arg", appName);
        return "RobotNaiveGui";
    }

    @PostMapping("/configure")
    public String configure(Model viewmodel  , @RequestParam String move, String addr ){
        ColorsOut.outappl("HIController | configure:" + move, ColorsOut.BLUE);
        ConnQakBase connToRobot = ConnQakBase.create( ProtocolType.tcp );
        conn = connToRobot.createConnection(addr, 8020);
        return "RobotNaiveGui";
    }

    @PostMapping("/robotmove")
    public String doMove(Model viewmodel  , @RequestParam String move ){
                     //String @RequestParam(name="move", required=false, defaultValue="h")robotmove  )  {
        //sysUtil.colorPrint("HIController | param-move:$robotmove ", Color.RED)
        ColorsOut.outappl("HIController | doMove:" + move + " robotName=" + robotName, ColorsOut.BLUE);
        if( move.equals("t")){
            Qak22Util.sendAMsg( SystemData.startSysCmd("hicontroller",robotName) );
        }else{
            try {
                String msg = moveAril(move).toString();
                ColorsOut.outappl("HIController | doMove msg:" + msg , ColorsOut.BLUE);
                conn.forward( msg );
            } catch (Exception e) {
                ColorsOut.outerr("HIController | doMove ERROR:"+e.getMessage());
                //e.printStackTrace();
            }
        }
        return "RobotNaiveGui";
    }

    @ExceptionHandler
    public ResponseEntity handle(Exception ex) {
        HttpHeaders responseHeaders = new HttpHeaders();
        return new ResponseEntity(
                "BaseController ERROR " + ex.getMessage(),
                responseHeaders, HttpStatus.CREATED);
    }
}

File: webBasicrobotqak/src/main/kotlin/it/unibo/webBasicrobotqak/HIController.kt
package it.unibo.webBasicrobotqak

import com.andreapivetta.kolor.Color
import connQak.ConnectionType
import connQak.connQakBase
import it.unibo.actor0.sysUtil
import it.unibo.kactor.MsgUtil
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.*
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.beans.factory.annotation.Autowired

/*

 */

@Controller
class HIController {
    @Value("\${human.logo}")
    var appName: String?    = null
    /*
     * Update the page vie socket.io when the application-resource changes.
     * See also https://www.baeldung.com/spring-websockets-send-message-to-user
     */
    @Autowired
    var  simpMessagingTemplate : SimpMessagingTemplate? = null

    //lateinit var connSupport     : connQakBase
    lateinit var coapsupport     : CoapSupport
    lateinit var connToRobot     : connQakBase
    lateinit var robotAddr       : String
    val channel      = Channel<String>()
    val coapObserver = WebPageCoapHandler(this, null)
    var moveresult="done"
    companion object{
        var logo = "just to test"
        //var answerMoveChannel  = Channel<String>()
        var robotanswerChannel  : Channel<String>? = null
        fun setAnswerChannel(answChannel  : Channel<String>){
            sysUtil.colorPrint("HIController | answChannelllllllllllll: ${answChannel}", Color.BLUE)
            robotanswerChannel = answChannel
        }

    }

    //var ws         = IssWsHttpJavaSupport.createForWs ("localhost:8083")
    fun configure(addr : String){
        connQak.robothostAddr = addr
        robotAddr             = addr
        if( addr == "localhost" ) {/*
            try {
                scopeTorunBasicrobot.launch {
                    QakContext.createContexts(
                        "localhost", this, "basicrobot.pl", "sysRules.pl"
                    )
                }
            } catch (e: Exception) {
                sysUtil.colorPrint("HIController | Error $e", Color.RED)
            }
            while (basicrobot == null) {
                Thread.sleep(500)
                val br = QakContext.getActor("basicrobot")
                if (br == null) {
                    sysUtil.colorPrint("HIController | waiting for basicrobot creation ...", Color.GREEN)
                } else basicrobot = br
            }*/
        }else { //basicrobotAddr != "localhost"
            //basicrobot = null
            connToRobot = connQakBase.create(ConnectionType.TCP)
            connToRobot.createConnection()
        }
        coapsupport =
            CoapSupport("coap://${connQak.robothostAddr}:${connQak.robotPort}",
                "ctxbasicrobot/basicrobot")
        coapsupport.observeResource( coapObserver ) //WebPageCoapHandler(this, channel)

        sysUtil.colorPrint("HIController | INIT $simpMessagingTemplate", Color.GREEN)
    }//configure



    @GetMapping("/")    //defines that the method handles GET requests.
    fun entry(viewmodel: Model): String {
         viewmodel.addAttribute("viewmodelarg", appName)
        sysUtil.colorPrint("HIController | entry model=$viewmodel", Color.GREEN)
        return  "basicrobotqakGui"
    }

    //http://localhost:8085/moveOnget?move=w  WARNING: not RESTful
    @GetMapping("/moveOnget")    //defines that the method handles GET requests.
    fun doMoveOnGet(viewmodel: Model,
        @RequestParam(name="move", required=false, defaultValue="h")robotmove : String
    ): String {
        viewmodel.addAttribute("viewmodelarg", robotmove)
        viewmodel.addAttribute("robotAddr",  robotAddr) //always ...
        sysUtil.colorPrint("HIController | doMoveOnGet=$robotmove", Color.RED)
        return  "basicrobotqakGui"
    }

    @PostMapping("/configure")
    fun handleConfigure(viewmodel : Model,
      @RequestParam(name="move", required=false, defaultValue="h")addr : String) : String {
        if( addr!="localhost"){
            configure(addr)
            viewmodel.addAttribute("viewmodelarg", "configured with basicrobot addr="+addr)
            viewmodel.addAttribute("robotAddr",  robotAddr)
        }else{
            viewmodel.addAttribute("viewmodelarg", "localhost not allowed")
        }

        return  "basicrobotqakGui"
    }


    //curl -i -d "move=w" -X POST http://localhost:8085/robotmove
    //it is a good practice to return the location of the newly created resource in the response header.
    @PostMapping("/robotmove")
    fun  doMove(viewmodel : Model,
        @RequestParam(name="move", required=false, defaultValue="h")robotmove : String) : String{
        sysUtil.colorPrint("HIController | param-move:$robotmove ", Color.RED)

         //viewmodel.addAttribute("viewmodelarg", "${robotmove}") //resRep.content
         //connSupport.forward( MsgUtil.buildDispatch("webgui", "cmd", "cmd($robotmove)", connQak.qakdestination))
        //sysUtil.colorPrint("HIController | xxxxxxxxx... ${answerChannel} ", Color.RED)
        if( robotmove == "p" ){
            val reqMsg = MsgUtil.buildRequest("basicrobotproxy", "step", "step(350)", connQak.qakdestination)
            /*
            Set the channel in coapObserver, so to receive the answer, to be shown on the page
             */
            coapObserver.channel = channel
            connToRobot.request(reqMsg)
            sysUtil.colorPrint("HIController | waits on channel ", Color.BLUE)
            runBlocking{
                moveresult = channel.receive()
                sysUtil.colorPrint("HIController | receives $moveresult", Color.BLUE)
                //Reset the channel in coapObserver, so to allow normal updating
                coapObserver.channel = null
            }
        }else { //normal move
            val cmdMsg = MsgUtil.buildDispatch("hicontroller", "cmd", "cmd($robotmove)", connQak.qakdestination)
            //if( basicrobot !=null )  scopeTorunBasicrobot.launch { basicrobot!!.autoMsg(cmdMsg) }    //basicrobot.actor.send(cmdMsg)
            //else
                connToRobot.forward( cmdMsg )
        }
        //
        viewmodel.addAttribute("viewmodelarg", "${robotmove}:${moveresult}")
        viewmodel.addAttribute("robotAddr",  robotAddr) //always ...
        moveresult = "done"
        //sysUtil.colorPrint("HIController | return the page after move ", Color.RED)
        return "basicrobotqakGui"
    }
/*
    @MessageMapping("/app/update")
    @SendTo("/topic/infodisplay")
    fun updateTheMap(@Payload message: String): ResourceRep {
        sysUtil.colorPrint("HIController | message=$message", Color.BLUE)
        return ResourceRep("todo")
    }
*/
}